services/logs/filter.go

```diff
@@ -0,0 +1,216 @@
+package logs
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/onflow/flow-evm-gateway/storage"
+	"math/big"
+	"slices"
+)
+
+type FilterCriteria struct {
+	Addresses []common.Address
+	Topics    [][]common.Hash
+}
+
+// todo think about creating an interface for all the filters
+// Filter interface { Match() (chan *gethTypes.Log, error) }
+
+// RangeFilter matches all the indexed logs within the range defined as
+// start and end block height. The start must be strictly smaller than end value.
+type RangeFilter struct {
+	start, end *big.Int
+	criteria   FilterCriteria
+	receipts   storage.ReceiptIndexer
+}
+
+func NewRangeFilter(
+	start, end big.Int,
+	criteria FilterCriteria,
+	receipts storage.ReceiptIndexer,
+) *RangeFilter {
+	return &RangeFilter{
+		start:    &start,
+		end:      &end,
+		criteria: criteria,
+		receipts: receipts,
+	}
+}
+
+func (r *RangeFilter) Match() ([]*gethTypes.Log, error) {
+	if r.start.Cmp(r.end) != -1 {
```

Comment:
nit: this is a more go conventional way to express this
```suggestion
	if r.start.Cmp(r.end) >= 0 {
```